README
NAME
    Message::Passing::STOMP - input and output messages to STOMP.

SYNOPSIS
        # Terminal 1:
        $ message-pass --input STDIN --output STOMP --output_options \
            '{"destination":"/queue/foo","hostname":"localhost","port":"6163","username":"guest","password":"guest"}'
        {"data":{"some":"data"},"@metadata":"value"}

        # Terminal 2:
        $ message-pass --output STDOUT --input STOMP --input_options \
            '{"destination":"/queue/foo","hostname":"localhost","port":"6163","username":"guest","password":"guest"}'
        {"data":{"some":"data"},"@metadata":"value"}

DESCRIPTION
    A AnyEvent::STOMP transport for Message::Passing.

    This transport is currently alpha quality. It is known working, however
    it does not gracefully handle errors or disconnects from the stomp
    server.

SEE ALSO
    Message::Passing::Output::STOMP
    Message::Passing::Input::STOMP
    Message::Passing
    AnyEvent::STOMP

AUTHOR
    Tomas (t0m) Doran <bobtfish@bobtfish.net>

COPYRIGHT
    Copyright The above mentioned AUTHOR 2012.

LICENSE
    GNU Affero General Public License, Version 3

    If you feel this is too restrictive to be able to use this software,
    please talk to us as we'd be willing to consider re-licensing under less
    restrictive terms.



Changes
  - Fix checking AnyEvent::STOMP version to be done in a way PAUSE
    hopefully won't object to.

  - Internal refactoring to use roles common to other components.

  - Support subscribing to multiple queues or topics with a single
    Message::Passing::Input::STOMP

0.005
  - Better documentation in the Input and Output classes.

  - Drop dependencies on JSON modules.

  - Link to STOMP protocol documentation

  - Link to RabbitMQ and ActiveMQ as the recommended queues.

  - Document the semantics for STOMP queue names as implemented
    by RabbitMQ and ActiveMQ.

0.004
   - Work with changes in Message::Passing 0.006

0.003
   - Fix to use the correct ->_set_connected(0) API in the
     connection manager, so that upcoming reconnect features
     work as expected.

0.002
   - Fix input and output to be able to connect to a configureable
     queue or topic.

   - Fix documentation to be vagely useful

   - Note code is currently alpha quality

0.001_01
    - Initial version for smoking



Makefile.PL
use strict;
use warnings;
use inc::Module::Install;
use Module::Install::Authority;

all_from 'lib/Message/Passing/STOMP.pm';

license 'AGPL_3';
authority('GHENRY');
resources(
    repository => "git://github.com/suretec/Message-Passing-STOMP.git",
);

requires 'Moose';
requires 'namespace::autoclean';
requires 'AnyEvent';
requires 'AnyEvent::STOMP' => '0.6';
requires 'Try::Tiny';
requires 'Task::Weaken';
requires 'Message::Passing' => '0.008';

test_requires 'Test::More' => '0.88';

author_requires(
    'Test::Pod' => '1.14',
    'Test::NoTabs' => '0',
    'Test::Pod::Coverage' => '1.04',
    'Pod::Coverage' => '0.19',
    'Pod::Coverage::TrustPod' => '0',
    'Test::Spelling' => '0',
    'Net::Stomp' => '0',
    'JSON' => '2.00',
);

author_tests 't/author';

if ($Module::Install::AUTHOR) {
    system("pod2text lib/Message/Passing/STOMP.pm > README")
        and die $!;
}

WriteAll;



lib/Message/Passing/STOMP/ConnectionManager.pm
package Message::Passing::STOMP::ConnectionManager;
use Moose;
use Scalar::Util qw/ weaken /;
use AnyEvent;
use AnyEvent::STOMP;
use Carp qw/ croak /;
use namespace::autoclean;

BEGIN { # For RabbitMQ https://rt.cpan.org/Ticket/Display.html?id=68432
    if ($AnyEvent::STOMP::VERSION <= 0.6) {
        no warnings 'redefine';
        sub AnyEvent::STOMP::send_frame {
            my $self = shift;
            my ($command, $body, $headers) = @_;

            croak 'Missing command' unless $command;

            $headers->{'content-length'} = length $body || 0;
            $body = '' unless defined $body;

            my $frame = sprintf("%s\n%s\n\n%s\000",
                        $command,
                        join("\n", map { "$_:$headers->{$_}" } keys %$headers),
                        $body);
            $self->{handle}->push_write($frame);
        }
    }
}

with qw/
    Message::Passing::Role::ConnectionManager
    Message::Passing::Role::HasHostnameAndPort
/;

sub _default_port { 6163 }

has ssl => (
    is => 'ro',
    isa => 'Bool',
    default => 0,
);

has [qw/ username password /] => (
    is => 'ro',
    isa => 'Str',
    default => 'guest',
);

sub _build_connection {
    my $self = shift;
    weaken($self);
    my $client = AnyEvent::STOMP->connect(
        $self->hostname, $self->port, $self->ssl, undef, 0,
        {
            'accept-version' => '1.1',
            host => '/',
            login => $self->username,
            passcode => $self->password,
        },
        {},
    );
    $client->reg_cb(CONNECTED => -2000 => sub {
        my ($client, $handle, $host, $port, $retry) = @_;
        $self->_set_connected(1);
    });
    $client->reg_cb(io_error => sub {
        my ($client, $errmsg) = @_;
        warn("IO ERROR $errmsg");
        $self->_set_connected(0);
    });
    $client->reg_cb(connect_error =>  sub {
        my ($client, $errmsg) = @_;
        warn("CONNECT ERROR $errmsg");
        $self->_set_connected(0);
    });
    return $client;
}

__PACKAGE__->meta->make_immutable;
1;

=head1 NAME

Message::Passing::STOMP::ConnectionManager - Wraps an AnyEvent::STOMP connection.

=head1 ATTRIBUTES


=head1 AUTHOR, COPYRIGHT AND LICENSE

See L<Message::Passing::STOMP>.

=cut



lib/Message/Passing/STOMP.pm
package Message::Passing::STOMP;
use Moose ();
use namespace::autoclean;

our $VERSION = "0.005";
$VERSION = eval $VERSION;

1;

=head1 NAME

Message::Passing::STOMP - input and output messages to STOMP.

=head1 SYNOPSIS

    # Terminal 1:
    $ message-pass --input STDIN --output STOMP --output_options \
        '{"destination":"/queue/foo","hostname":"localhost","port":"6163","username":"guest","password":"guest"}'
    {"data":{"some":"data"},"@metadata":"value"}

    # Terminal 2:
    $ message-pass --output STDOUT --input STOMP --input_options \
        '{"destination":"/queue/foo","hostname":"localhost","port":"6163","username":"guest","password":"guest"}'
    {"data":{"some":"data"},"@metadata":"value"}

=head1 DESCRIPTION

An L<STOMP|http://stomp.github.com/> transport for L<Message::Passing>.

L<STOMP|http://stomp.github.com/> is a simple and well supported messaging protocol, with clients in
many languages.

This is a simple adaptor to allow you to send or receive STOMP messages

=head1 QUEUES AND TOPICS

STOMP does not specify the semantics for message destinations or subscriptions (other than
that they are a string), however most common STOMP servers (including ActiveMQ and RabbitMQ)
expect a string matching /queue/name or /topic/name.

Messages published to a destination prefixed by /queue/ are queued up, until a subscriber connects,
at which point they are delivered. If multiple subscribers connect then the messages are distributed
between the subscribers (i.e. each subscriber sees a portion of the messages).

Messages published to a destination prefixed by /topic/ are copied to all connected subscribers,
however if there are no subscribers currently, they are discarded, rather than being queued up.

=head1 STOMP SERVERS

There are a number of message brokers which will communicate with STOMP clients, the main ones
being:

=head2 RabbitMQ

L<RabbitMQ|http://www.rabbitmq.com/> has L<a STOMP plugin|http://www.rabbitmq.com/stomp.html>.

=head2 ActiveMQ

L<ActiveMQ|http://activemq.apache.org/> has L<a STOMP plugin|http://activemq.apache.org/stomp.html>.

=head1 SEE ALSO

=over

=item L<http://stomp.github.com/> - STOMP protocol documentation

=item L<Message::Passing::Output::STOMP>

=item L<Message::Passing::Input::STOMP>

=item L<Message::Passing>

=item L<AnyEvent::STOMP>

=back

=head1 AUTHOR

Tomas (t0m) Doran <bobtfish@bobtfish.net>

=head1 COPYRIGHT

Copyright The above mentioned AUTHOR 2012.

=head1 LICENSE

GNU Affero General Public License, Version 3

If you feel this is too restrictive to be able to use this software,
please talk to us as we'd be willing to consider re-licensing under
less restrictive terms.

=cut

1;

